File: LR_1/src/git/group/Serialization.java
package git.group;

import git.group.Builder.Builder;
import git.group.List.TList;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.Objects;

public class Serialization {
    public  static void saveToFile(TList list, String filename, String type) throws Exception {
        try (PrintWriter writer = new PrintWriter(filename)) {
            writer.println(type);
            list.forEach((C)->writer.println(C + " "));
        }
        System.out.println("\nСписок сохранен!");
    }

    public static TList loadFile(String filename) throws Exception {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String type = br.readLine();
            if (!Factory.getTypeNameList().contains(type)) {
                throw new Exception("Неправильная файловая структура");
            }
            String line = br.readLine();
            String[] items = line.split(" ");
            Builder[] arr = new Builder[items.length];

            for (int i = 0; i < arr.length; i++)
                if (!Objects.equals(items[i], "null"))
                    arr[i] = (Builder) Factory.getBuilderByName(type).parseObject(items[i]);

            System.out.println("\nСписок был загружен!");
            return new TList(arr);
        }
    }
}

File: LR_3/src/git/group/Serialization.kt
package git.group

import git.group.Builder.Builder
import git.group.List.DoIt
import git.group.List.TList
import java.io.BufferedReader
import java.io.FileReader
import java.io.PrintWriter

object Serialization {
    @Throws(Exception::class)
    fun saveToFile(list: TList, filename: String?, type: String?) {
        PrintWriter(filename).use { writer ->
            writer.println(type)
            list.forEach(object : DoIt {
                override fun doIt(o: Any?) {
                    writer.print(o)
                }
            })
        }
        println("\nСписок не сохранен!")
    }

    @Throws(Exception::class)
    fun loadFile(filename: String?): TList {
        BufferedReader(FileReader(filename)).use { br ->
            val type = br.readLine()
            if (!Factory.typeNameList.contains(type)) {
                throw Exception("Неправильная файловая структура")
            }
            val line = br.readLine()
            val items = line.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val arr = arrayOfNulls<Builder>(items.size)
            for (i in arr.indices) if (items[i] != "null") arr[i] = Factory.getBuilderByName(type)?.parseObject(
                items[i]
            ) as Builder
            println("\nСписок был загружен!")
            return TList(arr)
        }
    }
}